app/modules/enduser_setup.c

```diff
@@ -1333,27 +1347,84 @@ static err_t enduser_setup_http_recvcb(void *arg, struct tcp_pcb *http_client, s
     return ERR_ABRT;
   }
 
+  tcp_arg_t *l = arg; 
+
   if (!p) /* remote side closed, close our end too */
   {
     ENDUSER_SETUP_DEBUG("connection closed");
-    scan_listener_t *l = arg; /* if it's waiting for scan, we have a ptr here */
-    if (l)
-      remove_scan_listener (l);
+    if (l) {
+      if (l->magic == SCAN_LISTENER_MAGIC) {
+        remove_scan_listener ((scan_listener_t *)l);
+      } else if (l->magic == HTTP_REQUEST_BUFFER_MAGIC) {
+        free(l);
+      }
+    }
 
     deferred_close (http_client);
     return ERR_OK;
   }
 
-  char *data = calloc (1, p->tot_len + 1);
-  if (!data)
-    return ERR_MEM;
+  http_request_buffer_t *hrb;
+  if (!l) {
+    hrb = calloc(1, sizeof(*hrb));
+    if (!hrb) {
+      ENDUSER_SETUP_ERROR("http_recvcb failed. Unable to send HTML.", ENDUSER_SETUP_ERR_UNKOWN_ERROR, ENDUSER_SETUP_ERR_NONFATAL);
+    }
+    hrb->magic = HTTP_REQUEST_BUFFER_MAGIC;
+    tcp_arg(http_client, hrb);
+  } else if (l->magic == HTTP_REQUEST_BUFFER_MAGIC) {
+    hrb = (http_request_buffer_t *) l;
+  } else {
+    ENDUSER_SETUP_ERROR("http_recvcb failed. Unable to send HTML.", ENDUSER_SETUP_ERR_UNKOWN_ERROR, ENDUSER_SETUP_ERR_NONFATAL);
+  }
+
+  // Append the new data
+  size_t newlen = hrb->length + p->tot_len;
+  hrb = realloc(hrb, sizeof(*hrb) + newlen + 1);
+  tcp_arg(http_client, hrb);
+  if (!hrb) {
+    ENDUSER_SETUP_ERROR("http_recvcb failed. Unable to send HTML.", ENDUSER_SETUP_ERR_UNKOWN_ERROR, ENDUSER_SETUP_ERR_NONFATAL);
+  }
+
+  pbuf_copy_partial(p, hrb->data + hrb->length, p->tot_len, 0);
+  hrb->data[newlen] = 0;
+  hrb->length = newlen;
 
-  unsigned data_len = pbuf_copy_partial (p, data, p->tot_len, 0);
   tcp_recved (http_client, p->tot_len);
   pbuf_free (p);
 
+  // see if we have the whole request.
+  // Rely on the fact that the header shoyld not contain a null character
+  const char *end_of_header = strstr(hrb->data, "\r\n\r\n");
+  if (end_of_header == 0) {
+    return ERR_OK;
+  }
+
+  end_of_header += 4;
+
+  // We have the entire header, now see if there is any content
+  for (const char *hdr = hrb->data; hdr && hdr < end_of_header; hdr = strchr(hdr, '\n')) {
+    hdr += 1; // Skip the \n
+    if (strncasecmp(hdr, "Content-length", 14) == 0) {
+      // There is a content-length header
+      const char *field = hdr + 14;
+      while (*field != '\n' && *field != ':') {
+        field++;
+      }
+      if (*field == ':') {
+        size_t extra = strtol(field + 1, 0, 10);
+        if (extra + end_of_header - hrb->data > hrb->length) {
+          return ERR_OK;
+        }
+      }
+    }
+  }
```

Comment:
And if there isn't a Content-Length header?

Reply:
That seems fishy to me.  https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.4 seems, to my very naive mind, to suggest that we should be bailing with 411 Length Required because, well, we probably require a length if this isn't a GET request, or something like that.

Reply:
> we should be bailing with 411 Length Required because, well, we probably require a length if this isn't a GET request

I second that.

Reply:
The browser can always use chunked encoding which removes the need for the content-length header. I've now added code to handle that case. I *doubt* it will happen since the post parameters are small.

I didn't want to make the code that checked whether the entire request had arrived depend on the HTTP method being used. This way, it will correctly determine whether the entire request has been received. The individual handlers can then operate on the entire request.